SEGGER/SEGGER_RTT.c

```diff
@@ -1876,7 +1899,20 @@ int SEGGER_RTT_SetFlagsDownBuffer(unsigned BufferIndex, unsigned Flags) {
 *
 */
 void SEGGER_RTT_Init (void) {
+
+  /*
+   * If used by bootloader or by stand-alone application, the standard
+   * Segger code using _DoInit() is used. For applications started by
+   * bootloader, Segger recommends not calling this function, as all the
+   * recommended APIs use INIT(). Unfortunately Zephyr log backend uses
+   * SEGGER_RTT_WriteSkipNoLock() that doesn't call INT(), so we need

```

Comment:
typo: `INIT`

Reply:
Maybe we could add a dedicated Kconfig option to let user decide if Zephyr application should actually force overwrite of the RTT structures during initialization. The current approach cannot support e.g. other bootloaders.
That would also allow to preserve RTT content during e.g. software reboot that does not clear RAM content.

Reply:
I also wonder if it wouldn't be reasonable to check whole `acID` right before initialization (instead of checking just the first byte) to ensure no false positives in case RAM content is not zero-initialized (with current implementation chance of false positive when checking if structure is initialized is quite high - `~1/255`)

Reply:
If we are going to rely on bootloader to initialize the RTT structures, we need to ensure backwards compatibility of the initialization operation (many embedded applications rely on immutable bootloader that cannot be updated once flashed).

Reply:
Thanks for your suggestion. Unfortunately I was too busy and I'll be out now for a couple of week, but afterwards I'll try to follow them.
The Kconfig option is a good idea, I was not very satisfied with my #ifdefs that are less general. I just have to figure out where such a Kconfig can be placed.
I also agree on the whole `acID` check, the `INIT()` macro can be used like it is only for other functions than `SEGGER_RTT_Init()`. I tried not to touch the SEGGER module, then trying with minimal change, but this is indeed reasonable.
I'll even rebase to RTT 3.40.

Reply:
With latest release I tried to take account of all above suggestions:
 * Rebased to SystemView V340
 * Managed dedicated CONFIG_SEGGER_RTT_INIT_MODE to have different init behaviours

The second change gives more flexibility: application level configs can force initialization of control block without condition or have a complete or partial check on acID to have a contitional init. Yet this requires some more changes in Segger code.

The SystemView V340 code misses some minor changes (safer INIT(), const _aTerminalId) tht are provided by latest SystemView V352a: I've added them, as in previous version.

Reply:
Thanks for all your changes looks good :) I've done similar modifications but haven't pushed changes to Zephyr due to having it interop with other parts of downstream Nordic Fork, which I couldn't figure out how to handle in a nice way. However having this upstream would definitely be a good addition.